northstar/src/npk/npk.rs

```diff
@@ -748,6 +758,92 @@ fn sign_hashes(key_pair: &Keypair, hashes_yaml: &str) -> String {
+                    context_buf.len(),
+                )
+            };
+            if length == -1 {
+                return Err(Error::Selinux(format!(
+                    "Failed to read selinux context of {}.",
+                    &path.display(),
+                )));
+            }
+
+            if let Ok(old_context) = CStr::from_bytes_with_nul(&context_buf[..length as usize]) {
+                if let Ok(old_context) = old_context.to_str() {
+                    // Replace selinux type of current file
+                    let old_context = old_context.to_string();
+                    let mut components: Vec<_> = old_context.split(':').collect();
+                    if components.len() >= 3 {
+                        components[2] = context_type;
+                    } else {
+                        return Err(Error::Selinux(format!(
+                            "Unexpected selinux context format: {}",
+                            &old_context,
+                        )));
+                    }
+
+                    // Write updated context
+                    let new_context = components.join(":");
+                    if let Ok(new_context) = CString::new(new_context.as_bytes()) {
+                        if unsafe {
+                            nix::libc::setxattr(
+                                path_c.as_ptr() as *const nix::libc::c_char,
+                                name_c.as_ptr() as *const nix::libc::c_char,
+                                new_context.as_ptr() as *mut core::ffi::c_void,
+                                new_context.len(),
+                                nix::libc::XATTR_REPLACE,
+                            )
+                        } != 0
+                        {
+                            return Err(Error::Selinux(format!(
+                                "Failed to call setxattr on {}",
+                                &path.display()
+                            )));
+                        }
+                    } else {
+                        return Err(Error::Selinux(format!(
+                            "Failed to convert new selinux context to C string: {:x?}",
+                            &new_context.as_bytes()
+                        )));
+                    }
+                } else {
+                    return Err(Error::Selinux(format!(
+                        "Failed to convert old selinux context to string: {}",
+                        &path.display()
+                    )));
+                }
+            } else {
+                return Err(Error::Selinux(format!(
+                    "Failed to convert old selinux context to C string: {}",
+                    &path.display()
+                )));
+            }
+        } else {
+            return Err(Error::Selinux(format!(
+                "Failed to convert path to C string: {}",
+                &path.display()
+            )));
+        }
```

Comment:
Is it possible to flatten this by creating some bindings. This is really hard to read and the errors are very similar. Maybe move it to a fn and just use a single error message.

Reply:
I broke the steps down into sub-functions that only return one error each.